src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1628 @@
+        },
+        [&](Index i) {
+          return SignatureResultLocation{heapType, i};
+        });
+    }
+  }
+
+  // Creates a location for a null of a particular type and adds a root for it.
+  // Such roots are where the default value of an i32 local comes from, or the
+  // value in a ref.null.
+  Location getNullLocation(Type type) {
+    auto location = NullLocation{type};
+    addRoot(location, PossibleContents::literal(Literal::makeZero(type)));
+    return location;
+  }
+
+  // Iterates over a list of children and adds links from them. The target of
+  // those link is created using a function that is passed in, which receives
+  // the index of the child.
+  void handleChildList(ExpressionList& operands,
+                       std::function<Location(Index)> makeTarget) {
+    Index i = 0;
+    for (auto* operand : operands) {
+      // This helper is not used from places that allow a tuple (hence we can
+      // hardcode the index 0 a few lines down).
+      assert(!operand->type.isTuple());
+
+      if (isRelevant(operand->type)) {
+        info.links.push_back({ExpressionLocation{operand, 0}, makeTarget(i)});
+      }
+      i++;
+    }
+  }
+
+  void visitStructNew(StructNew* curr) {
+    if (curr->type == Type::unreachable) {
+      return;
+    }
+    auto type = curr->type.getHeapType();
+    if (curr->isWithDefault()) {
+      // Link the default values to the struct's fields.
+      auto& fields = type.getStruct().fields;
+      for (Index i = 0; i < fields.size(); i++) {
+        info.links.push_back(
+          {getNullLocation(fields[i].type), DataLocation{type, i}});
+      }
+    } else {
+      // Link the operands to the struct's fields.
+      handleChildList(curr->operands, [&](Index i) {
+        return DataLocation{type, i};
+      });
+    }
+    addRoot(curr, PossibleContents::exactType(curr->type));
+  }
+  void visitArrayNew(ArrayNew* curr) {
+    if (curr->type == Type::unreachable) {
+      return;
+    }
+    auto type = curr->type.getHeapType();
+    if (curr->init) {
+      info.links.push_back(
+        {ExpressionLocation{curr->init, 0}, DataLocation{type, 0}});
+    } else {
+      info.links.push_back(
+        {getNullLocation(type.getArray().element.type), DataLocation{type, 0}});
+    }
+    addRoot(curr, PossibleContents::exactType(curr->type));
+  }
+  void visitArrayInit(ArrayInit* curr) {
+    if (curr->type == Type::unreachable) {
+      return;
+    }
+    if (!curr->values.empty()) {
+      auto type = curr->type.getHeapType();
+      handleChildList(curr->values, [&](Index i) {
+        // The index i is ignored, as we do not track indexes in Arrays -
+        // everything is modeled as if at index 0.
+        return DataLocation{type, 0};
+      });
+    }
+    addRoot(curr, PossibleContents::exactType(curr->type));
+  }
+
+  // Struct operations access the struct fields' locations.
+  void visitStructGet(StructGet* curr) {
+    if (!isRelevant(curr->ref)) {
```

Comment:
Nominal or isorecursive! It would be good to start allowing isorecursive types in these optimizations because we should switch to them by default at some point.

Reply:
Can we find all the subtypes of an isorecursive type? Doesn't that depend on other isorecursive recursion groups etc.? I guess new groups can't be added if we assume a closed world, is that the idea?

Regardless, this depends on `SubTyping` which atm has an assertion that it only works in nominal typing, so I guess that should be changed as well? Not sure how we'd want to handle the closed world concept there, though, if that's relevant - it's not a pass, it's a utility (for a pass, we can imagine a simple mechanism some day to not run closed-world passes if a flag says the world is open).

Reply:
I think it's literally just a matter of fixing that assertion. Since the isorecursive types have declared supertypes, all the supertype-related APIs work identically between nominal and isorecursive types.

Reply:
Ok, sounds good, done.

I guess when we get to open-world situations we'll have a bunch more work on this stuff.